src/main/java/frc/robot/subsystems/drive/Drivetrain.java

```diff
@@ -1,16 +1,329 @@
+  }
+
+  /**
+   * Resets the odometry to the specified pose.
+   *
+   * @param pose The pose to which to set the odometry.
+   */
+  public void resetOdometry(Pose2d pose) {
+    m_odometry.resetPosition(
+        Rotation2d.fromRadians(-getGyroAngle().in(Units.Radians)),
+        new SwerveModulePosition[] {
+          m_frontLeft.getPosition(),
+          m_frontRight.getPosition(),
+          m_rearLeft.getPosition(),
+          m_rearRight.getPosition(),
+        },
+        pose);
+  }
+
+  /**
+   * drives the drivatrain using the given inputs the magnitude of the joystick components shouldn't
+   * be > 1 (x^2 + y^2 <= 1)
+   *
+   * @param xSpeed the x-pos of the left joystick (-1, 1)
+   * @param ySpeed the y-pos of the left joystick (-1, 1)
+   * @param xRot the x-pos of the right joystick (-1, 1)
+   * @param yRot the x-pos of the right joystick (-1, 1)
+   * @param altDrive whether or not to use the alternative turning mode
+   * @param centerGyro whether or not to reset the gyro position to the current rotation
+   */
+  public void drive(Vector spdVec, Vector rotVec, boolean altDrive, boolean centerGyro) {
+    if (centerGyro) zeroHeading();
+    if (altDrive) {
+      altDrive(spdVec, rotVec);
+    } else {
+      mainDrive(spdVec, rotVec.x());
+    }
+  }
+
+  /**
+   * moves the drivetrain using the main turning mode
+   *
+   * @param xSpeed the proportion of the robot's max velocity to move in the x direction
+   * @param ySpeed the proportion of the robot's max velocity to move in the y direction
+   * @param xRot the speed to rotate with (-1, 1)
+   */
+  private void mainDrive(Vector spdVec, double xRot) {
+    double rot = xRot * DriveConfig.kMaxAngularSpeed;
+    move(spdVec, rot);
+  }
+
+  /**
+   * gets the value of the robot's gyro as a Measure<Angle>
+   *
+   * @see Measure
+   * @return the angle of the robot gyro
+   */
+  private Measure<Angle> getGyroAngle() {
+    return Units.Degrees.of(m_gyro.getAngle());
+  }
+
+  /**
+   * moves the drivetrain using the alternative turning mode
+   *
+   * @param xSpeed the proportion of the robot's max velocity to move in the x direction
+   * @param ySpeed the proportion of the robot's max velocity to move in the y direction
+   * @param xRot the x component of the direction vector to point towards
+   * @param yRot the y component of the direction vector to point towards
+   */
+  private void altDrive(Vector spdVec, Vector rotVec) {
+    double rot = 0;
+    m_rightAngGoalRadians = rotVec.angle();
+    if (rotVec.squaredMag() > 0) {
+      double stickAng = m_rightAngGoalRadians;
+      // gets the difference in angle, then uses mod to make sure its from -PI rad to PI rad
+      rot = altTurnSmooth(stickAng);
+    }
+    m_turnDirRadians = rot;
+    move(spdVec, rot);
+  }
+
+  private double altTurnSmooth(double stickAng) {
+    return Math.tanh(
+            ((getGyroAngle().in(Units.Radians) + stickAng + Math.PI) % (2 * Math.PI) - Math.PI)
```

Comment:
Is stickAng in radians?

Reply:
It should be it seems. That should also be a `Measure<Angle>`